main/lsp/LSPTypechecker.h

```diff
@@ -61,27 +86,46 @@ class LSPTypechecker final {
     /** List of files that have had errors in last run*/
     std::vector<core::FileRef> filesThatHaveErrors;
     std::unique_ptr<KeyValueStore> kvstore; // always null for now.
+    /** Set only when typechecking is happening on the slow path. Contains all of the state needed to restore
+     * LSPTypechecker to its pre-slow-path state. */
+    std::optional<UndoState> cancellationUndoState;
 
     std::shared_ptr<const LSPConfiguration> config;
+    /** Used to preempt running slow paths. */
+    std::shared_ptr<core::lsp::PreemptionTaskManager> preemptManager;
     /** Used for assertions. Indicates if `initialize` has been run. */
     bool initialized = false;
 
     /** Conservatively reruns entire pipeline without caching any trees. Returns 'true' if committed, 'false' if
      * canceled. */
-    bool runSlowPath(LSPFileUpdates updates, WorkerPool &workers, bool cancelable);
+    bool runSlowPath(LSPFileUpdates updates, WorkerPool &workers, bool cancelable,
+                     std::optional<std::shared_ptr<core::lsp::PreemptionTaskManager>> preemptManager = std::nullopt);
```

Comment:
could we make `preemptManager` required for uniformity?

Reply:
You know what? I think I can. It'll currently be a no-op if `cancelable` is `false`.